libqtile/widget/touchpad.py

```diff
@@ -0,0 +1,144 @@
+    """
+    A simple widget to display touchpad state.
+
+    You can also bind keyboard shortcuts to the Touchpad widget with:
+
+    .. code-block:: python
+
+        Key(
+            [],
+            "XF86TouchpadToggle",
+            lazy.widget["touchpad"].toggle()
+        )
+    """
+
+    device_id: None | str
+    get_state_func: None | Callable[[str], bool]
+    set_state_func: None | Callable[[str, bool], None]
+
+    defaults: list[tuple[str, Any, str]] = [
+        (
+            "device_id",
+            None,
+            "Touchpad name or id, provided as ``str``. "
+            "If set to ``None`` the widget will try to find it automatically using ``xinput``.",
+        ),
+        (
+            "get_state_func",
+            None,
+            "Function which can read the touchpad state. "
+            "It should return the state as ``bool``. "
+            "If set to ``None`` a default function based on ``xinput`` will be used.",
+        ),
+        (
+            "set_state_func",
+            None,
+            "Function which can enable/disable touchpad. "
+            "It should take one ``bool`` parameter and change touchpad state accordingly. "
+            "If set to ``None`` a default function based on ``xinput`` will be used.",
+        ),
+        ("format", "{state}", "Displayed text format"),
+        ("enabled_char", "[👆]", "State shown when touchpad is enabled"),
+        ("disabled_char", "[🚫]", "State shown when touchpad is disabled"),
+        (
+            "update_interval",
+            10,
+            "Seconds between status updates. "
+            "Additionally, status is updated immediately after calling ``toggle`` command from this class.",
+        ),
+    ]
+
+    def __init__(self, **config) -> None:
+        base.ThreadPoolText.__init__(self, "", **config)
+        self.add_defaults(Touchpad.defaults)
+
+        self.add_callbacks(
+            {
+                "Button1": self.toggle,
+            }
+        )
+
+    def _configure(self, qtile, bar) -> None:
+        base.ThreadPoolText._configure(self, qtile, bar)
+
+    def poll(self) -> str:
+        if self.device_id is None:
+            device_id = self._get_touchpad_device_name()
+        if self.get_state_func is None:
+            get_state_func = self._get_touchpad_enabled
+
+        state = get_state_func(device_id)
+        state_text = self.enabled_char if state else self.disabled_char
+        return self.format.format(state=state_text)
+
+    @expose_command()
+    def toggle(self) -> None:
+        """Toggle touchpad on/off."""
+
+        if self.device_id is None:
+            device_id = self._get_touchpad_device_name()
+        if self.get_state_func is None:
+            get_state_func = self._get_touchpad_enabled
+        if self.set_state_func is None:
+            set_state_func = self._set_touchpad_enabled
+
+        new_state = not get_state_func(device_id)
+        set_state_func(device_id, new_state)
```

Comment:
You check `self.get_state_func` twice. Once here, and once in `poll`. You could just do it once in `_configure`.

Also, if `self.get_state_func` is not `None` then `get_state_func` is undefined which means line 118 would fail.

Same issue for `set_state_func`.

Reply:
Right, thanks for spotting that

Reply:
Done. I also had to add some type asserts to keep mypy happy